src/leap/mail/sync_hooks.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+# sync_hooks.py
+# Copyright (C) 2015 LEAP
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from re import compile as regex_compile
+
+from zope.interface import implements
+from twisted.internet import defer
+from twisted.plugin import IPlugin
+from twisted.python import log
+
+from leap.soledad.client.interfaces import ISoledadSyncedPlugin
+from leap.mail import constants
+
+_get_doc_type_preffix = lambda s: s[:2]
+
+
+class MailProcessingPostSyncHook(object):
+    implements(IPlugin, ISoledadSyncedPlugin)
+
+    META_DOC_PREFFIX = _get_doc_type_preffix(constants.METAMSGID)
+    watched_doc_types = (META_DOC_PREFFIX, )
+
+    _account = None
+    _pending_docs = []
+    _processing_deferreds = []
+
+    def process_received_docs(self, doc_id_list):
+        if self._has_configured_account():
+            process_fun = self._make_uid_index
+        else:
+            self._processing_deferreds = []
+            process_fun = self._queue_doc_id
+
+        for doc_id in doc_id_list:
+            if _get_doc_type_preffix(doc_id) in self.watched_doc_types:
+                log.msg("Mail post-sync hook: processing %s" % doc_id)
+                process_fun(doc_id)
+
+        if self._processing_deferreds:
+            return defer.gatherResults(self._processing_deferreds)
```

Comment:
This has not being addressed. I'm ok with the code as it is if you like it as it is, but I like more to remove the two ifs here and in `_process_queued_docs` I find clearer methods that always return the same type of value.